tests/fixtures/testModel.ts

```diff
@@ -66,6 +66,12 @@ export interface TestModel extends Model {
   TypeAliasCase2?: TypeAliasModelCase2;
 }
 
+export interface UnionTestModel {
+  or: TypeAliasModel1 | TypeAliasModel2;
+  and: TypeAliasModel1 & TypeAliasModel2;
+  referenceAnd: TypeAliasModelCase1;
+}
+
```

Comment:
There are a couple of other cases that users have asked for in the past in tsoa issues.

1) I'd like to see a case here where users try to create an "on-the-fly" union.

So given this interface:
```ts
export interface OnTheFlyUnion {
  twoObjects: { apple: string } |  { orange: string }
}
```
And we'll need some input from other people (@endor?) on what we should be testing for. As a first pass, we'll need to probably throw or warn that you must use interfaces.

Perhaps as a future enhancement we could create a name on the fly, but I don't know if we should start down the road of generating swagger model names for users. Like what would this one even be... `$ref: twoObjectsObjectWithAppleOrObjectWithOrange`? Because that won't scale very well if objects have lots of properties. Perhaps the easiest way would be to just create a UUID since that's unlikely to be a duplicate $ref. So like `$ref: #/definitions/Union913ced83-3873-43bc-ab5f-6a7c22af9181`

Another case that we'll want to test for since people have asked for it:
```ts
export interface IObjectWithUnionOfPrimativeAndObj {
  primOrObj : { apple: string } | number
}
```

and lastly:
```ts
export interface IObjectWithUnionOfNullAndObj {
  primOrNull : null | { apple: string }
}
```

Reply:
Due to the way these things are being handled, all of those cases should not be an issue.
I simply collect the types that are mixed (& or |) and resolve each one individually.
I added some tests to illustrate the behavior. :beers:

Reply:
The `primOrNull` will throw (`Error: Unknown type: NullKeyword`) because null individually can not be resolved.
I'm not sure it should be, maybe there is an argument to catch that corner case.

Reply:
> Perhaps as a future enhancement we could create a name on the fly, but I don't know if we should start down the road of generating swagger model names for users. Like what would this one even be... $ref: twoObjectsObjectWithAppleOrObjectWithOrange? Because that won't scale very well if objects have lots of properties. Perhaps the easiest way would be to just create a UUID since that's unlikely to be a duplicate $ref. So like $ref: #/definitions/Union913ced83-3873-43bc-ab5f-6a7c22af9181

This is also related to https://github.com/lukeautry/tsoa/pull/400#discussion_r313035787.

> collect the types that are mixed (& or |) and resolve each one individually

This PR will only $ref what it used to $ref individually before and inline what it used to inline before.
If this is a problem, and I don't believe it is (when we $ref, the user has a corresponding interface in his code, which lines up elegantly with the openapi schema), it's something that should be tackled for all of tsoa, not just unions/intersection types.

Reply:
> The primOrNull will throw (Error: Unknown type: NullKeyword) because null individually can not be resolved.
I'm not sure it should be, maybe there is an argument to catch that corner case.

I think in OpenAPI 3.0 we could check if any of the union types are the null keyword, and if so we could mark it as nullable.

Reply:
> I think in OpenAPI 3.0 we could check if any of the union types are the null keyword and if so, we could mark it as nullable.

I thought about this and while it sounds nice on paper and is easy to detect, it will be a very big change under the hood because validation will break. There, we currently do not distinguish between `required` and `nullable` properties, so I'd have to add that, check which version we are validating against and ideally add the distinction between `nullable` and `required` for all data types. The tests also seem to rely on the fact that `x-nullable` and `required` are somewhat the same concept for Swagger 2.

Therefore, I think this is more than a convenience and users should rely on the `?` operator for now.
Since `null` throws, everyone knows what's going on so communication works.

We can always follow up with a validation for OAPI3 that does not break and depends on a RouteGenerator that distinguishes between both concepts.

Reply:
> Therefore, I think this (T|null) is more than a convenience and users should rely on the ? operator for now.
I mean it might be an issue for some since null is not undefined. But I'll mark this as resolved because

> Since null throws, everyone knows what's going on so communication works.

Yes, I suppose that's true. I'm sure that some user of tsoa will complain, but we can always enhance this code when/if that user ever files a PR.